File: orchestrator/apps/intercompany_pools/kvo18_advance_vat_offset_preview.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import Any

from .kvo18_advance_vat_offset_intake import (
    ADVANCE_INVOICE_KVO01_SLOT,
    ADVANCE_OFFSET_KVO18_SLOT,
    CASH_RECEIPT_ORDER_SLOT,
    DECLARATION_EVIDENCE_SLOT,
    KVO18_ADVANCE_VAT_OFFSET_POLICY_REVISION,
    Kvo18AdvanceVatOffsetBatchNormalizationResult,
    Kvo18AdvanceVatOffsetIntakeRow,
    build_kvo18_advance_vat_offset_policy_config,
)


KVO18_DUPLICATE_ROW_ID = "KVO18_DUPLICATE_ROW_ID"
KVO18_DUPLICATE_ROW_FINGERPRINT = "KVO18_DUPLICATE_ROW_FINGERPRINT"


@dataclass(frozen=True)
class Kvo18AdvanceVatOffsetPreview:
    policy_revision: str
    total_amount: str
    total_vat_amount: str
    currency: str
    row_count: int
    rows: list[dict[str, Any]]
    stages: dict[str, dict[str, Any]]
    technical_realization_policy: dict[str, Any]
    evidence_requirements: dict[str, Any]
    diagnostics: list[dict[str, Any]]
    content_hash: str

    def as_dict(self) -> dict[str, Any]:
        return {
            "policy_revision": self.policy_revision,
            "total_amount": self.total_amount,
            "total_vat_amount": self.total_vat_amount,
            "currency": self.currency,
            "row_count": self.row_count,
            "rows": list(self.rows),
            "stages": dict(self.stages),
            "technical_realization_policy": dict(self.technical_realization_policy),
            "evidence_requirements": dict(self.evidence_requirements),
            "diagnostics": list(self.diagnostics),
            "content_hash": self.content_hash,
        }


def build_kvo18_advance_vat_offset_preview(
    *,
    batch: Kvo18AdvanceVatOffsetBatchNormalizationResult,
) -> Kvo18AdvanceVatOffsetPreview:
    policy = build_kvo18_advance_vat_offset_policy_config()
    total_amount = sum((row.amount for row in batch.rows), Decimal("0.00"))
    total_vat_amount = sum((row.vat_amount for row in batch.rows), Decimal("0.00"))
    currency = batch.rows[0].currency if batch.rows else "RUB"
    return Kvo18AdvanceVatOffsetPreview(
        policy_revision=KVO18_ADVANCE_VAT_OFFSET_POLICY_REVISION,
        total_amount=str(total_amount.quantize(Decimal("0.01"))),
        total_vat_amount=str(total_vat_amount.quantize(Decimal("0.01"))),
        currency=currency,
        row_count=len(batch.rows),
        rows=[_preview_row(row=row) for row in batch.rows],
        stages=_build_stage_preview(),
        technical_realization_policy=dict(policy["technical_realization"]),
        evidence_requirements=dict(policy["declaration_evidence"]),
        diagnostics=_build_duplicate_diagnostics(rows=batch.rows),
        content_hash=batch.content_hash,
    )


def _preview_row(*, row: Kvo18AdvanceVatOffsetIntakeRow) -> dict[str, Any]:
    return {
        "line_no": row.line_no,
        "row_id": row.row_id,
        "row_fingerprint": row.row_fingerprint,
        "counterparty": {
            "ref": row.counterparty_ref,
            "name": row.counterparty_name,
        },
        "contract": {
            "ref": row.contract_ref,
            "name": row.contract_name,
        },
        "operation_date": row.operation_date.isoformat(),
        "amount": str(row.amount),
        "vat_rate": row.vat_rate,
        "vat_amount": str(row.vat_amount),
        "currency": row.currency,
        "source_reference": row.source_reference,
        "source_document_number": row.source_document_number,
        "lineage": {
            "cash_receipt_order": {"required": True, "state": "pending"},
            "advance_invoice_kvo01": {"required": True, "state": "blocked"},
            "advance_offset_kvo18": {"required": True, "state": "blocked"},
            "declaration_evidence": {"required": True, "state": "blocked"},
        },
    }


def _build_stage_preview() -> dict[str, dict[str, Any]]:
    return {
        CASH_RECEIPT_ORDER_SLOT: {
            "slot_key": CASH_RECEIPT_ORDER_SLOT,
            "label": "Создать ПКО",
            "state": "ready",
            "prerequisites": [],
            "produces": ["cash_receipt_order_ref"],
        },
        ADVANCE_INVOICE_KVO01_SLOT: {
            "slot_key": ADVANCE_INVOICE_KVO01_SLOT,
            "label": "Создать СФ на аванс",
            "state": "blocked",
            "prerequisites": [CASH_RECEIPT_ORDER_SLOT],
            "produces": ["advance_invoice_ref", "sales_book_kvo01_evidence"],
        },
        ADVANCE_OFFSET_KVO18_SLOT: {
            "slot_key": ADVANCE_OFFSET_KVO18_SLOT,
            "label": "Сформировать зачет (КВО 18)",
            "state": "blocked",
            "prerequisites": [ADVANCE_INVOICE_KVO01_SLOT, "operator_preview_confirmation"],
            "produces": [
                "technical_realization_ref",
                "technical_realization_final_state",
                "purchase_book_kvo18_evidence",
            ],
        },
        DECLARATION_EVIDENCE_SLOT: {
            "slot_key": DECLARATION_EVIDENCE_SLOT,
            "label": "Проверить декларацию",
            "state": "blocked",
            "prerequisites": [ADVANCE_OFFSET_KVO18_SLOT],
            "produces": ["declaration_projection_kvo01_kvo18"],
        },
    }


def _build_duplicate_diagnostics(
    *,
    rows: list[Kvo18AdvanceVatOffsetIntakeRow],
) -> list[dict[str, Any]]:
    diagnostics: list[dict[str, Any]] = []
    diagnostics.extend(
        _duplicate_diagnostics_for_key(
            rows=rows,
            key_name="row_id",
            code=KVO18_DUPLICATE_ROW_ID,
            values=[row.row_id for row in rows],
        )
    )
    diagnostics.extend(
        _duplicate_diagnostics_for_key(
            rows=rows,
            key_name="row_fingerprint",
            code=KVO18_DUPLICATE_ROW_FINGERPRINT,
            values=[row.row_fingerprint for row in rows],
        )
    )
    return diagnostics


def _duplicate_diagnostics_for_key(
    *,
    rows: list[Kvo18AdvanceVatOffsetIntakeRow],
    key_name: str,
    code: str,
    values: list[str],
) -> list[dict[str, Any]]:
    line_numbers_by_value: dict[str, list[int]] = {}
    for row, value in zip(rows, values, strict=True):
        line_numbers_by_value.setdefault(value, []).append(row.line_no)

    diagnostics: list[dict[str, Any]] = []
    for value, line_numbers in sorted(line_numbers_by_value.items()):
        if len(line_numbers) < 2:
            continue
        diagnostics.append(
            {
                "code": code,
                "severity": "error",
                "field": key_name,
                "value": value,
                "line_numbers": line_numbers,
                "detail": f"Duplicate KVO18 advance VAT offset source {key_name}.",
            }
        )
    return diagnostics


__all__ = [
    "KVO18_DUPLICATE_ROW_FINGERPRINT",
    "KVO18_DUPLICATE_ROW_ID",
    "Kvo18AdvanceVatOffsetPreview",
    "build_kvo18_advance_vat_offset_preview",
]

File: frontend/src/pages/Pools/kvo18AdvanceVatOffsetPreview.ts
export const KVO18_ADVANCE_VAT_OFFSET_SCHEME_CODE = 'kvo18-advance-vat-offset'
export const KVO18_ADVANCE_VAT_OFFSET_SCHEMA_TEMPLATE_CODE = 'kvo18-advance-vat-offset-intake'
export const KVO18_ADVANCE_VAT_OFFSET_POLICY_REVISION = 'kvo18_advance_vat_offset_policy.v1'

export const CASH_RECEIPT_ORDER_SLOT = 'cash_receipt_order'
export const ADVANCE_INVOICE_KVO01_SLOT = 'advance_invoice_kvo01'
export const ADVANCE_OFFSET_KVO18_SLOT = 'advance_offset_kvo18'
export const DECLARATION_EVIDENCE_SLOT = 'declaration_evidence'

export type Kvo18StageSlot =
  | typeof CASH_RECEIPT_ORDER_SLOT
  | typeof ADVANCE_INVOICE_KVO01_SLOT
  | typeof ADVANCE_OFFSET_KVO18_SLOT
  | typeof DECLARATION_EVIDENCE_SLOT

export type Kvo18AdvanceVatOffsetStage = {
  slotKey: Kvo18StageSlot
  label: string
  state: 'ready' | 'blocked'
  prerequisites: string[]
  produces: string[]
}

export type Kvo18AdvanceVatOffsetDiagnostic = {
  code: string
  field: string
  value: string
  lineNumbers: number[]
  detail: string
}

export type Kvo18AdvanceVatOffsetPreview = {
  policyRevision: string
  totalAmount: string
  totalVatAmount: string
  currency: string
  rowCount: number
  stages: Record<Kvo18StageSlot, Kvo18AdvanceVatOffsetStage>
  technicalRealizationPolicy: {
    documentStateAfterOffset: string
    confirmationGates: string[]
  }
  evidenceRequirements: {
    salesBookKvo01Required: boolean
    purchaseBookKvo18Required: boolean
    declarationProjectionRequired: boolean
    documentCreationAloneIsSuccess: boolean
  }
  diagnostics: Kvo18AdvanceVatOffsetDiagnostic[]
}

type NormalizedRow = {
  lineNo: number
  counterpartyRef: string
  counterpartyName: string
  contractRef: string
  contractName: string
  operationDate: string
  amount: number
  vatRate: string
  vatAmount: number
  currency: string
  rowId: string
  fingerprintKey: string
}

export function isKvo18AdvanceVatOffsetSchemaTemplate(template: {
  code: string
  metadata?: Record<string, unknown>
} | null | undefined): boolean {
  if (!template) {
    return false
  }
  return template.code === KVO18_ADVANCE_VAT_OFFSET_SCHEMA_TEMPLATE_CODE
    || template.metadata?.scheme_code === KVO18_ADVANCE_VAT_OFFSET_SCHEME_CODE
}

export function buildKvo18AdvanceVatOffsetPreview(payload: unknown): Kvo18AdvanceVatOffsetPreview {
  const rows = normalizeRows(payload)
  const totalAmount = rows.reduce((sum, row) => sum + row.amount, 0)
  const totalVatAmount = rows.reduce((sum, row) => sum + row.vatAmount, 0)
  return {
    policyRevision: KVO18_ADVANCE_VAT_OFFSET_POLICY_REVISION,
    totalAmount: formatAmount(totalAmount),
    totalVatAmount: formatAmount(totalVatAmount),
    currency: rows[0]?.currency ?? 'RUB',
    rowCount: rows.length,
    stages: buildStages(),
    technicalRealizationPolicy: {
      documentStateAfterOffset: 'unposted_after_purchase_book_evidence',
      confirmationGates: [
        'operator_preview_confirmed',
        'cash_receipt_order_evidence',
        'advance_invoice_kvo01_evidence',
      ],
    },
    evidenceRequirements: {
      salesBookKvo01Required: true,
      purchaseBookKvo18Required: true,
      declarationProjectionRequired: true,
      documentCreationAloneIsSuccess: false,
    },
    diagnostics: buildDuplicateRowDiagnostics(rows),
  }
}

function normalizeRows(payload: unknown): NormalizedRow[] {
  const rawRows = Array.isArray(payload)
    ? payload
    : (isRecord(payload) && Array.isArray(payload.rows) ? payload.rows : null)
  if (!rawRows || rawRows.length === 0) {
    throw new Error('rows must be a non-empty list')
  }
  return rawRows.map((row, index) => {
    if (!isRecord(row)) {
      throw new Error(`row ${index + 1}: object expected`)
    }
    const lineNo = index + 1
    const currency = requiredText(row, 'currency', lineNo).toUpperCase()
    if (currency !== 'RUB') {
      throw new Error(`row ${lineNo}: unsupported currency ${currency}`)
    }
    const counterpartyRef = requiredText(row, 'counterparty_ref', lineNo)
    const contractRef = requiredText(row, 'contract_ref', lineNo)
    const operationDate = requiredText(row, 'operation_date', lineNo)
    const rowId = requiredText(row, 'row_id', lineNo)
    const amount = requiredAmount(row, 'amount', lineNo)
    const vatAmount = requiredAmount(row, 'vat_amount', lineNo)
    const vatRate = requiredText(row, 'vat_rate', lineNo)
    return {
      lineNo,
      counterpartyRef,
      counterpartyName: requiredText(row, 'counterparty_name', lineNo),
      contractRef,
      contractName: requiredText(row, 'contract_name', lineNo),
      operationDate,
      amount,
      vatRate,
      vatAmount,
      currency,
      rowId,
      fingerprintKey: [
        counterpartyRef,
        contractRef,
        operationDate,
        rowId,
        formatAmount(amount),
        vatRate,
        formatAmount(vatAmount),
        currency,
      ].join('|'),
    }
  })
}

function buildStages(): Record<Kvo18StageSlot, Kvo18AdvanceVatOffsetStage> {
  return {
    [CASH_RECEIPT_ORDER_SLOT]: {
      slotKey: CASH_RECEIPT_ORDER_SLOT,
      label: 'Создать ПКО',
      state: 'ready',
      prerequisites: [],
      produces: ['cash_receipt_order_ref'],
    },
    [ADVANCE_INVOICE_KVO01_SLOT]: {
      slotKey: ADVANCE_INVOICE_KVO01_SLOT,
      label: 'Создать СФ на аванс',
      state: 'blocked',
      prerequisites: [CASH_RECEIPT_ORDER_SLOT],
      produces: ['advance_invoice_ref', 'sales_book_kvo01_evidence'],
    },
    [ADVANCE_OFFSET_KVO18_SLOT]: {
      slotKey: ADVANCE_OFFSET_KVO18_SLOT,
      label: 'Сформировать зачет (КВО 18)',
      state: 'blocked',
      prerequisites: [ADVANCE_INVOICE_KVO01_SLOT, 'operator_preview_confirmation'],
      produces: ['technical_realization_ref', 'technical_realization_final_state', 'purchase_book_kvo18_evidence'],
    },
    [DECLARATION_EVIDENCE_SLOT]: {
      slotKey: DECLARATION_EVIDENCE_SLOT,
      label: 'Проверить декларацию',
      state: 'blocked',
      prerequisites: [ADVANCE_OFFSET_KVO18_SLOT],
      produces: ['declaration_projection_kvo01_kvo18'],
    },
  }
}

function buildDuplicateRowDiagnostics(rows: NormalizedRow[]): Kvo18AdvanceVatOffsetDiagnostic[] {
  const lineNumbersByRowId = new Map<string, number[]>()
  for (const row of rows) {
    lineNumbersByRowId.set(row.rowId, [...(lineNumbersByRowId.get(row.rowId) ?? []), row.lineNo])
  }
  return [...lineNumbersByRowId.entries()]
    .filter(([, lineNumbers]) => lineNumbers.length > 1)
    .map(([value, lineNumbers]) => ({
      code: 'KVO18_DUPLICATE_ROW_ID',
      field: 'row_id',
      value,
      lineNumbers,
      detail: 'Duplicate KVO18 advance VAT offset row_id.',
    }))
}

function requiredText(row: Record<string, unknown>, key: string, lineNo: number): string {
  const value = String(row[key] ?? '').trim()
  if (!value) {
    throw new Error(`row ${lineNo}: ${key} required`)
  }
  return value
}

function requiredAmount(row: Record<string, unknown>, key: string, lineNo: number): number {
  const value = Number.parseFloat(String(row[key] ?? '').trim().replace(',', '.'))
  if (!Number.isFinite(value)) {
    throw new Error(`row ${lineNo}: ${key} required`)
  }
  return value
}

function formatAmount(value: number): string {
  return value.toFixed(2)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}
